src/kbc_modules/eaa_kbc/mod.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::kbc_modules::{KbcCheckInfo, KbcInterface};
+use anyhow::*;
+use log::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::net::TcpStream;
+use std::os::unix::io::AsRawFd;
+
+pub mod protocol;
+pub mod rats_tls;
+use protocol::*;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct AnnotationPacket {
+    pub kid: String,
+    pub wrapped_data: Vec<u8>,
+    pub iv: Vec<u8>,
+    pub algorithm: String,
+    pub key_length: u16,
+}
+
+pub struct EAAKbc {
+    pub kbs_uri: String,
+    pub protocol_version: String,
+    pub algorithm: String,
+    pub key_length: u16,
+    // pub kek_cache: HashMap<String, String>,
+    pub tcp_stream: Option<TcpStream>,
+    pub tls_handle: Option<rats_tls::RatsTls>,
+}
+
+impl KbcInterface for EAAKbc {
+    fn check(&self) -> Result<KbcCheckInfo> {
+        let mut kbs_info: HashMap<String, String> = HashMap::new();
+        kbs_info.insert("kbs_addr".to_string(), self.kbs_uri.clone());
+        kbs_info.insert(
+            "protocol_version".to_string(),
+            self.protocol_version.clone(),
+        );
+        Ok(KbcCheckInfo { kbs_info: kbs_info })
+    }
+
+    fn decrypt_payload(&mut self, annotation: &str) -> Result<Vec<u8>> {
+        debug!("EAA KBC decrypt_payload() is called!");
+        let annotation_packet: AnnotationPacket = serde_json::from_str(annotation)?;
+        self.algorithm = annotation_packet.algorithm;
+        self.key_length = annotation_packet.key_length;
+        if self.tcp_stream.is_none() {
+            debug!("First request, connecting KBS...");
+            self.establish_new_kbs_connection()?;
+            debug!("connect success! TLS is established!");
+        }
+
+        debug!("start decrypt...");
+        let decrypted_payload = self.kbs_decrypt_payload(
+            annotation_packet.wrapped_data,
+            annotation_packet.kid,
+            annotation_packet.iv,
+        )?;
+        debug!("decrypted success!");
+        Ok(decrypted_payload)
+    }
+}
+
+impl EAAKbc {
+    pub fn new(kbs_uri: String) -> EAAKbc {
+        EAAKbc {
+            kbs_uri: kbs_uri,
+            protocol_version: String::new(),
+            algorithm: String::new(),
+            key_length: 0,
+            // kek_cache: HashMap::new(),
+            tcp_stream: None,
+            tls_handle: None,
+        }
+    }
+
+    fn establish_new_kbs_connection(&mut self) -> Result<()> {
+        debug!("create RATS TLS handle...");
+        self.tls_handle = Some(
+            rats_tls::RatsTls::new(
+                false,
+                0,
+                &Some("openssl".to_string()),
+                &Some("openssl".to_string()),
+                &Some("tdx".to_string()),
```

Comment:
So we're hard-coding this for TDX? It's a little bit difficult to tell what each of these parameters is and it seems mybe they should be configurable somewhere?

Reply:
These parameters are the configuration parameters defined by rats-tls. For the specific meanings of these parameters, please refer to the [rats-tls project](https://github.com/alibaba/inclavare-containers/tree/master/rats-tls).

At present, we really don't have a very suitable means to configure these parameters. In [issue#7](https://github.com/containers/attestation-agent/issues/6#issuecomment-924740146), @liangzhou121 proposes to extend the keyprovider protocol and pass these KBC specific parameters to AA by ocicrypt. These KBC specific parameters can be placed in pod annotation (or somewhere else) together with KBC name and KBS URI, after this is determined, the parameters here will become configurable, but at present, we can only do the demo by hard coding these parameters.

Reply:
ok. in the README this is described as a general-purpose KBC, but that doesn't fit with hard-coding TDX stuff. I guess maybe it is fine temporarily.

Do you think the KBC could automatically select the attester? For example could it check if is a TDX-enabled VM and then always use the TDX attester if it is?

In the other repo, we were discussing moving the KBC configuration to a pod annotation, which I think is a good idea in the long term. That would not work with some of these fields, however. We can't let the pod annotation select the attester or the verifier without compromising security (a rogue CSP could set verifier and attester to null). We need to be careful how we set the parameters. Doing it in the config would be secure, but it might be easier of the client if the KBC can set them automatically. 

A related question, why is the KBC selecting the verifier? Shouldn't it be up to the KBS to use whatever verifier it wants?

Reply:
It's a great idea to automatically detect the platform and select the attester instance. I have confirmed with the developers of rats-tls that rats-tls already has a development plan for this function, this is the [issue(#1427)](https://github.com/alibaba/inclavare-containers/issues/1427).

The verifier parameter here does not select the verifier instance of KBS. The rats-tls instance is symmetrical to the client side and the server side, the verifier parameter will only work when the `server` parameter is set to true. These parameters may cause some confusion, so as you suggested, I will add some comments on these parameters to explain.